commands/fetch_templates.go

```diff
@@ -18,113 +18,160 @@ import (
-		relativePath := strings.Replace(z.Name, "faas-cli-master/", "", -1)
-		if strings.Index(relativePath, "template") == 0 {
-			fmt.Printf("Found \"%s\"\n", relativePath)
-			rc, err := z.Open()
-			if err != nil {
-				return err
-			}
+		var rc io.ReadCloser
 
-			err = createPath(relativePath, z.Mode())
-			if err != nil {
-				return err
-			}
+		relativePath := z.Name[strings.Index(z.Name, "/")+1:]
+		if strings.Index(relativePath, "template/") != 0 {
+			// Process only directories inside "template" at root
+			continue
+		}
+
+		var language string
+		if languageSplit := strings.Split(relativePath, "/"); len(languageSplit) > 2 {
+			language = languageSplit[1]
+
+			if len(languageSplit) == 3 && relativePath[len(relativePath)-1:] == "/" {
+				// template/language/
+
+				if !canWriteLanguage(language, overwrite) {
+					existingLanguages = append(existingLanguages, language)
+					continue
+				}
+				countFetchedTemplates++
+			} else {
+				// template/language/*
 
-			// If relativePath is just a directory, then skip expanding it.
-			if len(relativePath) > 1 && relativePath[len(relativePath)-1:] != string(os.PathSeparator) {
-				err = writeFile(rc, z.UncompressedSize64, relativePath, z.Mode())
-				if err != nil {
-					return err
+				if !canWriteLanguage(language, overwrite) {
+					continue
 				}
 			}
+		} else {
+			// template/
+			continue
+		}
+
+		if rc, err = z.Open(); err != nil {
+			break
 		}
+
+		if err = createPath(relativePath, z.Mode()); err != nil {
+			break
+		}
+
+		// If relativePath is just a directory, then skip expanding it.
+		if len(relativePath) > 1 && relativePath[len(relativePath)-1:] != "/" {
+			if err = writeFile(rc, z.UncompressedSize64, relativePath, z.Mode()); err != nil {
+				break
+			}
+		}
+	}
+
+	if len(existingLanguages) > 0 {
+		log.Printf("Cannot overwrite the following (%d) directories: %v\n", len(existingLanguages), existingLanguages)
 	}
 
+	zipFile.Close()
+
+	log.Printf("Fetched %d template(s) from %s\n", countFetchedTemplates, templateURL)
+
+	err = removeArchive(archive)
+
+	return err
+}
+
+func removeArchive(archive string) error {
 	log.Printf("Cleaning up zip file...")
-	if _, err := os.Stat(ZipFileName); err == nil {
-		os.Remove(ZipFileName)
+	if _, err := os.Stat(archive); err == nil {
+		return os.Remove(archive)
 	} else {
 		return err
 	}
-	fmt.Println("")
-
-	return err
 }
 
-func fetchMasterZip(templateUrl string) error {
+// fetchMasterZip downloads a zip file from a repository URL
+func fetchMasterZip(templateURL string) (string, error) {
 	var err error
-	if _, err = os.Stat(ZipFileName); err != nil {
 
-		if len(templateUrl) == 0 {
-			templateUrl = "https://github.com/openfaas/faas-cli/archive/" + ZipFileName
-		}
+	templateURL = strings.TrimRight(templateURL, "/")
+	templateURL = templateURL + "/archive/master.zip"
 
+	templateURLSplit := strings.Split(templateURL, "/")
+	templateURLSplit = templateURLSplit[len(templateURLSplit)-4 : len(templateURLSplit)-2]
```

Comment:
I don't like the magic numbers -4 -2

What's this doing? maybe break out a function and document the indexes

Reply:
This is just to name the `master.zip` file as the name of the repository. It's not necessary so I'm removing it.